### DISCUSSION TITLE
Nested contravariant generics

### DISCUSSION THREAD

**Discussion by CoskunSunali** [dadhi/DryIoc#404] [Q&A] [answered]
Is it possible to satisfy the following requirement using DryIoc? I am trying to implement a simple event handler where A is base of B and anything related to B should trigger A as well. This is like what https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/OpenGenerics.md#generic-variance-when-resolving-many-services has as an example but the example there is covariant with the `out` keyword. However, event handlers usually are not expected to be covariant, they are rather contravariant because the event is passed in, not out.

E.g.: `void HandleEvent(T eventData)`

```cs
public static class DryIocContainer_GenericVarianceTest
{
    public static void Run()
    {
        var c = new Container();

        c.Register<IGenericType<A>, GenericType<A>>();
        c.Register<IGenericType<B>, GenericType<B>>();

        c.Register<IGenericTypeWrapper<IGenericType<A>>, GenericTypeWrapper<IGenericType<A>>>();
        c.Register<IGenericTypeWrapper<IGenericType<B>>, GenericTypeWrapper<IGenericType<B>>>();

        var genericTypes_A = c.ResolveMany<IGenericType<A>>(); // Gets A and B (expected)
        var genericTypes_B = c.ResolveMany<IGenericType<B>>(); // Gets B (expected)

        var genericTypeWrappers_A = c.ResolveMany<IGenericTypeWrapper<IGenericType<A>>>(); // Gets A (unexpected)
        var genericTypeWrappers_B = c.ResolveMany<IGenericTypeWrapper<IGenericType<B>>>(); // Gets A and B (unexpected)
    }
}

public interface IGenericTypeWrapper<in T> where T : class
{
}

public interface IGenericType<in T> where T : class
{
    void Do(T value);
}

public class GenericTypeWrapper<T> : IGenericTypeWrapper<T> where T : class
{
}

public class GenericType<T> : IGenericType<T> where T : class
{
    public void Do(T value)
    {
    }
}

public class A : IGenericType<A>
{
    public virtual void Do(A value)
    {
        // Do A
    }
}

public class B : A
{
    public override void Do(A value)
    {
        // Do B
    }
}
```

**Comment by dadhi** [comment, marked as answer]
Interesting question, will look

**Reply by CoskunSunali** [reply]
Thanks for your reply. Please don’t waste your time before I post a better example. I see something wrong with my requirements. I’m on mobile and cannot fix it, will be in front of my computer in an hour, will fix it and post an update. I’ll provide a more concrete example with a real world scenario.

**Reply by CoskunSunali** [reply]
Okay, the example was completely wrong in terms of "expected" and "unexpected" parts I defined.

I defined both interfaces as covariant but the inner generic had to be contravariant. The following code satisfies everything. DryIoc handles it correctly! Great job, @dadhi . I will mark your reply as an answer so that the question is marked as Answered.

```cs
    public static class DryIocContainer_GenericVarianceTest2
    {
        public static void Run()
        {
            var c = new Container();

            c.Register<IGenericTypeWrapper<IGenericType<A>>, GenericTypeWrapper<IGenericType<A>>>();
            c.Register<IGenericTypeWrapper<IGenericType<B>>, GenericTypeWrapper<IGenericType<B>>>();

            var genericTypeWrappers_A = c.ResolveMany<IGenericTypeWrapper<IGenericType<A>>>(); // Gets A (expected)
            var genericTypeWrappers_B = c.ResolveMany<IGenericTypeWrapper<IGenericType<B>>>(); // Gets A and B (expected)
        }
    }

    public interface IGenericTypeWrapper<in T> where T : class
    {
        void Do(T value);
    }

    public interface IGenericType<out T> where T : class
    {
    }

    public class GenericTypeWrapper<T> : IGenericTypeWrapper<T> where T : class
    {
        public void Do(T value)
        {
        }
    }

    public class GenericType<T> : IGenericType<T> where T : class
    {

    }

    public class A : IGenericType<A>
    {
    }

    public class B : A
    {
    }
```

**Reply by dadhi** [reply]
Great to hear.

There is also a non-default `WithVariantGenericTypesInResolve()` rule to enable variance for the single service resolution.

Btw, if you interested in the MediatR alike variant request/handlers you may look inzo DryIoc.Messages namespace https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/DryIoc.Messages.md

**Reply by CoskunSunali** [reply]
> There is also a non-default WithVariantGenericTypesInResolve() rule to enable variance for the single service resolution.

Is there an example of it somewhere in the docs? I did my best to find one but no luck so far and it is hard to understand what it helps with :)

> Btw, if you interested in the MediatR alike variant request/handlers you may look inzo DryIoc.Messages namespace https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/DryIoc.Messages.md

Thanks, I will take a look at it. Appreciated.

**Reply by dadhi** [reply]
> Is there an example of it somewhere in the docs?

Here are the tests for the related issue https://github.com/dadhi/DryIoc/blob/master/test/DryIoc.IssuesTests/GHIssue352_Consider_resolving_the_variance_compatible_open_generic_the_same_as_for_collection_of_open_generics.cs

**Reply by CoskunSunali** [reply]
Thanks @dadhi .

**Comment by CoskunSunali** [comment]
Hey @dadhi I have another question regarding open generics so thought it might be a better idea to continue within the same discussion. Just let me know if it's better for you if I create a new discussion.

My question is whether it is funny to try to do something like the following with DryIoc :)

```
Services.AddSingleton(
                typeof(ISuperWrapper<,>).MakeGenericType(typeof(IWrappedType<>), typeof(IResult)),
                typeof(SuperWrapper<>));
```

I know it is super ridiculous because `IWrappedType<>` and `SuperWrapper<>` would not even know which type parameter to use because type parameter counts are not equal but you thought so many magical things while developing DryIoc, I thought you might have an idea or just say don't even try it :)

**Reply by CoskunSunali** [reply]
What I am trying to achieve can be simplified as (even though it won't compile this way, just for making it easier to see what I am trying) :

```
Services.AddSingleton(
                typeof(ISuperWrapper<IWrappedType<>, IResult>),
                typeof(SuperWrapper<>));
```

So resolving `ISuperWrapper<A, IResult>` would return `SuperWrapper<A>`. `IResult` is useful here for creating different instances of `SuperWrapper<A>` based on different registrations.

**Reply by dadhi** [reply]
@CoskunSunali From the logic standpoint it makes sense.
Did you try the first code with `MakeGenericType`? Cause DryIoc has a  very sophisticated generic type matching and does not rely (almost) on the C# compiler in that regard.

**Reply by CoskunSunali** [reply]
Glad it makes sense to you.

I tried. I get the following exception:

```
ContainerException: code: Error.RegisteringNotAGenericTypedefServiceType;
message: Unsupported registration of service ISuperWrapper<IWrappedType<>, IResult> which is not a generic type definition but contains generic parameters.
Consider to register generic type definition ISuperWrapper<,> instead.
```

**Reply by dadhi** [reply]
Huh, ok. It may be too strict of the check. I will try to lift it and check the result  Tomorrow.

**Reply by CoskunSunali** [reply]
Thank you! Looking forward for testing it.

**Reply by dadhi** [reply]
The issue to track #406

**Reply by dadhi** [reply]
@CoskunSunali I have released DryIoc v4.8.0 with the fix.

**Reply by CoskunSunali** [reply]
It works! I will test a few scenarios and let you know how it goes. Thanks for the prompt reaction!
